### DISCUSSION TITLE
ReturnTypeWillChangeRector for JsonSerializable

### DISCUSSION THREAD

**Discussion by TobiasBg** [rectorphp/rector#7739] [Q&A] [answered]
Hi!

Similarly to https://github.com/rectorphp/rector/issues/6886, I want this behavior for `JsonSerializable`, but can't make it work :-(

I have this PHP class:
```PHP
<?php

class SomeClass implements JsonSerializable
{
	public function jsonSerialize(): mixed
	{
	}
}
```

I want to downgrade this to the PHP 7.2 (so that the `: mixed` return type is removed), but at the same time don't want deprecation warnings on newer versions of PHP, i.e. I want `#[\ReturnTypeWillChange]` to be added.

I'm trying with this rector.php:
```PHP
<?php

declare(strict_types=1);

use Rector\Config\RectorConfig;
use Rector\Set\ValueObject\DowngradeLevelSetList;
use Rector\Transform\Rector\ClassMethod\ReturnTypeWillChangeRector;
use Rector\Transform\ValueObject\ClassMethodReference;

return static function (RectorConfig $rectorConfig): void {
	$rectorConfig->paths([
		__DIR__ . '/src',
	]);

	$rectorConfig->sets( array(
		DowngradeLevelSetList::DOWN_TO_PHP_72,
	) );

	$rectorConfig->ruleWithConfiguration(
		ReturnTypeWillChangeRector::class,
		[new ClassMethodReference('JsonSerializable', 'jsonSerialize')]
	);
};
```

The downgrading part is working fine, but I can't get it to add the `#[\ReturnTypeWillChange]` attribute to the method :-(

I would be really grateful for any hints!

Thanks!
Tobias

**Comment by samsonasik** [comment, marked as answer]
You need to define `phpVersion()` for that:

```php
$rectorConfig->phpVersion(\Rector\Core\ValueObject\PhpVersion::PHP_81);
```

Ref https://getrector.com/demo/6051ff60-a502-493b-88e0-5e12f3e087de

**Reply by TobiasBg** [reply]
Hi @samsonasik,

ah, perfect! Yes, that makes sense. Thanks a lot!

Would it make sense to have this (regarding `JsonSerializable`/`jsonSerialize()`) built-in, similar to the `ArrayAccess` interface? Should I open an issue for that?

Best wishes,
Tobias

**Reply by samsonasik** [reply]
@TobiasBg yes, I am not looking into detailed for it yet, but that make sense for more general downgrade rules, so the Attribute need to be kept when downgrading, as they are marked as comment even not supported yet.

**Reply by samsonasik** [reply]
@TobiasBg Attribute seems actually kept when downgrading, if it already exists before, see https://getrector.com/demo/ab3ff020-d697-4b04-97db-4f8fe180f009, 

so, for `JsonSerializable` usage, to up before not exists, that require check php version first as the usage is depends of the user :)

**Reply by TobiasBg** [reply]
@samsonasik: Yes, it's kept, but I think it should automatically be added (without having to explicitly specify it in the rule), if e.g. `public function jsonSerialize(): mixed` is converted down to PHP 7.2. In my understanding, downgrading allows code to be run with the lower PHP version, but at the same time should not cause issue when it's run with current/new versions of PHP.
Or am I maybe misunderstanding this?

**Reply by samsonasik** [reply]
The `#[\ReturnTypeWillChange]` is depends of how user will add the return type or not, I will show the use case for you, in `kahlan`, we decide to not use `#[\ReturnTypeWillChange]` and directly apply the return type, see https://github.com/kahlan/kahlan/pull/392#issuecomment-990988304
